### ISSUE TITLE
filter by attribute doensn't work

### ISSUE THREAD

**Issue by touch1111** [woocommerce/woocommerce#27519]
Hi,
I try to figure out why the filter by attribute doens't work.
I have set attribute on PRODUCT-> ATTRIBUTES with some TERMS.
Added two different TERMS in 3 products (just to try it ) and add the widget in side woocomerce bar, but the filter isn't showed...

I even tried to install the storefront template in a staging area, but doesn't work.

Can you help me ?

**Comment by mathiasdb** [comment]
Hi

I'm having a similar issue. Our client gave us the feedback that some filters were missing from the sidebar. After looking into the issue it only applies to Filter Products by Attribute. Feel free to check out the [website](https://www.topstore-fashion.be/product-categorie/work-fashion/).

After some digging in the code if found that the widgets are rendered by the WC_Widget_Layered_Nav class. Before outputting the widget, the current Products Query is "analysed", so the correct filters are showing for the current product "selection". The function responsible for this is get_filtered_term_product_counts(). 

Now the "real" issue starts here, on line 362 there's an extra rule that excludes **Variable Products** from the results.
https://github.com/woocommerce/woocommerce/blob/master/includes/widgets/class-wc-widget-layered-nav.php#L362

When I change line 362, just for testing purposes, into `$exclude_variable_products_tax_query_sql = $this->get_extra_tax_query_sql( 'product_type', array(), 'NOT IN' );`

All widgets of type Filter Products by Attributes will show up. Also note that the file and line 362 in question was changed on 28 July 2020. @see this [commit](https://github.com/woocommerce/woocommerce/commit/9de1306c21600aa11423361719f260c7069073cd)

It would be great if a solution or re-evaluation could be presented. Maybe the original author can help us out? @Konamiman

**Comment by Konamiman** [comment]
Hi @touch1111 @mathiasdb. Excluding variable products from the main query is on purpose, because later on the same function there's a second query that counts variable products (that one was introduced in https://github.com/woocommerce/woocommerce/pull/26260), but there are issues regarding products having "Any..." as the value of the attribute (see https://github.com/woocommerce/woocommerce/issues/27419). I'm working on a fix for that.

**Comment by mathiasdb** [comment]
@Konamiman thanks for the feedback!

**Comment by touch1111** [comment]
@Konamiman Thanks ! Hope you will fix soon 

If you need my login credentials can give you for testing

**Comment by treatsunleashed** [comment]
@Konamiman Thank you for addressing this.

I am having the same issue.  Started when I updated from 4.1 to 4.4.1.
Variable products are being excluded from the Filter Products by Attribute widget in the sidebar.  I'm using Brand as the attribute.

**Comment by bashgeek** [comment]
Have the same issue since the update to 4.4 (I think, very recent).

Getting the following SQL error with it:
`WordPress database error Unknown column 'wp_term_relationships.term_taxonomy_id' in 'where clause' for query SELECT COUNT( DISTINCT wp_posts.ID ) as term_count, terms.term_id as term_count_id FROM wp_posts
			INNER JOIN wp_term_relationships AS tr ON wp_posts.ID = tr.object_id
			INNER JOIN wp_term_taxonomy AS term_taxonomy USING( term_taxonomy_id )
			INNER JOIN wp_terms AS terms USING( term_id )

			WHERE
			wp_posts.post_status = 'publish'
			 AND (
  wp_posts.ID NOT IN (
				SELECT object_id
				FROM wp_term_relationships
				WHERE term_taxonomy_id IN (146)
			)
)
			AND (
				(
					wp_posts.post_type = 'product'
					 AND (
  wp_posts.ID NOT IN (
				SELECT object_id
				FROM wp_term_relationships
				WHERE term_taxonomy_id IN (4)
			)
)
					 AND (
  wp_term_relationships.term_taxonomy_id IN (70)
)
				)

			OR (wp_posts.post_type = 'product_variation'
		    AND NOT EXISTS (
		        SELECT ID FROM wp_posts AS parent
		        WHERE parent.ID = wp_posts.post_parent AND parent.post_status NOT IN ('publish')
		    ))

			)
			AND terms.term_id IN (131,132,44,67,38,228,37,94,275)
			 GROUP BY terms.term_id made by require('wp-blog-header.php'), require_once('wp-includes/template-loader.php'), include('/themes/neighborhood/woocommerce/archive-product.php'), dynamic_sidebar, WP_Widget->display_callback, WC_Widget_Layered_Nav->widget, WC_Widget_Layered_Nav->layered_nav_list, WC_Widget_Layered_Nav->get_filtered_term_product_counts`

This part is wrong:
`AND (
  wp_term_relationships.term_taxonomy_id IN (70)
)`

Should be:
`AND (
  **tr.**term_taxonomy_id IN (70)
)`

I hope this helps.

**Comment by adhamoc** [comment]
Hi there,

same issue with me, just updated from 4.2.0 to 4.4.1 and noticed that I can't filter variable products by attributes anymore and I'm actively using this option.

Hope you find a fix as soon as possible.

**Comment by grigorijurasov** [comment]
I can replicate with variable products.

Steps to replicate:

1. Create a variable product.
2. Add a "Filter Product by Attribute" widget https://d.pr/i/EjZQLs with an attribute this variable product is using.
3. Add the product to a product category and navigate to the category page where the widget should show up.
4. The widget is not displayed. 
5. Change the product type to "simple product" and the widget will show up.

3295008-zen

**Comment by TheGranadilla** [comment]
Oof, why this went live :S

**Comment by pgamaster** [comment]
please fix it asap, clients can't filter products by size

**Comment by mixalisen** [comment]
Hi guys,
just installed the new 4.5.0 version and the problem still exists.

**Comment by RobertVerdonschot** [comment]
This broke very important functionality in my clients store.
Good to see the original author is aware of the problem and is working it!

Is there anything we can do to help or give this issue more priority or attention?
(Status is still open and no labels assigned yet)

**Comment by pgamaster** [comment]
> 
> 
> This broke very important functionality in my clients store.
> Good to see the original author is aware of the problem and is working it!
> 
> Is there anything we can do to help or give this issue more priority or attention?
> (Status is still open and no labels assigned yet)

Use 3rd party filters as a temporary solution.

**Comment by Konamiman** [comment]
Hi all, please take a look at this: https://github.com/woocommerce/woocommerce/issues/27419#issuecomment-689657524

**Comment by RobertVerdonschot** [comment]
Thanks, I think it is a wise decision. Good luck.

**Comment by juliaamosova** [comment]
Hi everyone! 

We released [WooCommerce 4.5.2](https://developer.woocommerce.com/2020/09/14/woocommerce-4-5-2-fix-release/) just now - it contains the fix for the issue with the `Filter by Attribute` widget. Essentially, we rolled back to the way it used to work in WooCommerce 4.3. 

Please update to WooCommerce 4.5.2 and let us know if you are still experiencing any issues. Thanks for testing and for your patience everyone!

**Comment by rembo1992** [comment]
Awesome it's working again!

**Comment by adhamoc** [comment]
Yep, it works.

Thanks

**Comment by juliaamosova** [comment]
Thanks for testing and reporting that the issue had been resolved for you @rembo1992 and @adhamoc! At this point, I am closing this thread. However, please continue leaving comments in case anyone experienced any issues.

**Comment by amman20** [comment]
I have Woocommerce 4.8 and still experiencing the same problem

**Comment by digitalw** [comment]
The issue seems to still be in WC 5.2.0 any news on this?

**Comment by mithurshan92** [comment]
Wordpress 5.7.1 and the issue is still there. Not fixed. These people just want to close these threads but not fix them. Im done with wordpress moving to shopify

**Comment by ahmad-ali14** [comment]
Version 5.3.0 | still not working

**Comment by Acquamarina5** [comment]
Hi,

I’m new to WooCommerce and working on a brand new install, brand new site, using WooCommerce Version 5.4.1, and have the same problem. None of the filters work, with or without a plugin. They never worked from the start so I tried 3 different filter plugins after the original WooCommerce install thinking that a plugin was needed, creating a new widget for each one.

At first, I thought it was the fact that we had just migrated the content and products and the attributes were messed up. I fixed those and it still didn’t work.

Nothing works, nothing happens when a filter is selected or the page just hangs on loading using no plugins. On top of that, all the original sidebar widgets lost their styling from the premium theme once I placed them back on the sidebar. There seems to be a problem with styling CSS too on the sidebar as none of the widgets I created for the different filter plugins using Elementor Pro aligned properly. They look great on the preview but not when published. This happens regardless of the theme used.

I disabled all the installed plugins on the site and tried the default filters with the Storefront theme and the filters didn’t work either.

The search occasionally returns random empty pages when I know there are products that fit the query. It’s random and hard to reproduce.

There is no previous version to revert to since this is a fresh install. I don’t know what else to do. Any suggestions are greatly appreciated.
